tools/approval-service/internal/approvalservice/eventprocessor/githubprocessor/githubhandler.go

```diff
@@ -0,0 +1,249 @@
+}
+
+// Small interface to allow for easier testing of the Teleport client.
+// This is a subset of the teleport.Client interface that we need for our purposes.
+// It is not intended to be a complete representation of the Teleport API or the teleport.Client implementation.
+type teleClient interface {
+	CreateAccessRequestV2(ctx context.Context, req types.AccessRequest) (types.AccessRequest, error)
+	GetAccessRequests(ctx context.Context, filter types.AccessRequestFilter) ([]types.AccessRequest, error)
+}
+
+var _ accessrequest.ReviewHandler = &Processor{}
+
+// Opt is a function that modifies the GitHubHandler.
+type Opt func(r *Processor) error
+
+// WithLogger sets the logger for the GitHubHandler.
+func WithLogger(log *slog.Logger) Opt {
+	return func(r *Processor) error {
+		r.log = log
+		return nil
+	}
+}
+
+var defaultOpts = []Opt{
+	WithLogger(slog.Default()),
+}
+
+// New creates a new Processor.
+func New(ctx context.Context, cfg config.GitHubSource, tele teleClient, store store.GitHubService, opts ...Opt) (*Processor, error) {
+	p := &Processor{
+		store:      store,
+		teleClient: tele,
+		log:        slog.Default(),
+		org:        cfg.Org,
+		repo:       cfg.Repo,
+		envs:       cfg.Environments,
+	}
+
+	for _, o := range append(defaultOpts, opts...) {
+		if err := o(p); err != nil {
+			return nil, fmt.Errorf("applying option: %w", err)
+		}
+	}
+
+	f, err := os.Open(cfg.Authentication.App.PrivateKeyPath)
+	if err != nil {
+		return nil, fmt.Errorf("opening private key: %w", err)
+	}
+	defer func() {
+		err = errors.Join(err, f.Close())
+	}()
+	pKey, err := io.ReadAll(f)
+	if err != nil {
+		return nil, fmt.Errorf("reading private key: %w", err)
+	}
+
+	client, err := github.NewForApp(ctx, cfg.Authentication.App.AppID, cfg.Authentication.App.InstallationID, pKey)
+	p.client = client
+
+	return p, nil
+}
+
+// FindExistingAccessRequest checks if an Access Request already exists for the given GitHub deployment review event.
+func (p *Processor) FindExistingAccessRequest(ctx context.Context, e githubevents.DeploymentReviewEvent) (types.AccessRequest, error) {
+	list, err := p.teleClient.GetAccessRequests(ctx, types.AccessRequestFilter{})
+	if err != nil {
+		return nil, fmt.Errorf("getting access requests: %w", err)
+	}
+	for _, req := range list {
+		labels := req.GetStaticLabels()
+		if labels == nil {
+			continue
+		}
+		if labels["organization"] == e.Organization && labels["repository"] == e.Repository && labels["environment"] == e.Environment && labels["workflow_run_id"] == strconv.Itoa(int(e.WorkflowID)) {
+			return req, nil
+		}
+	}
+
+	// No existing access request found.
+	return nil, nil
+}
+
+// CreateNewAccessRequest creates a new Access Request for the given GitHub deployment review event.
+func (p *Processor) CreateNewAccessRequest(ctx context.Context, e githubevents.DeploymentReviewEvent, req types.AccessRequest) (types.AccessRequest, error) {
```

Comment:
nit: CreateNew.. is a bit verbose. I expect all things created are new at the time of creation. Perhaps just `CreateAccessRequest`.